Add a phase to Dotty that hides Dotty bottom types from JVM.

DarkDimius opened the pull request:
It is very convenient to have bottom types in type system. Unfortunately
    JVM does not have bottom types and we need to insert casts that would make
    bytecode pass verification.

DarkDimius commented:
/rebuild

DarkDimius commented:
@odersky please review.

smarter commented on src/dotty/tools/dotc/transform/Erasure.scala:
```diff
@@ -263,6 +263,12 @@ object Erasure extends TypeTestsCasts{
       }
+
+
+    /** Is `tpe` a type which is a bottom type for Dotty, but not a bottom type for JVM */
```
There is no bottom type on the JVM so I don't really understand the naming of this function, how is it different from `Definitions#isBottomType`?

smarter commented on src/dotty/tools/backend/jvm/BottomTypes.scala:
```diff
@@ -0,0 +1,81 @@
+  override def transformApply(tree: tpd.Apply)(implicit ctx: Context, info: TransformerInfo): tpd.Tree = {
+    val fun = tree.fun
+    val newArgs: List[tpd.Tree] = tree.args.zip(fun.tpe.dealias.firstParamTypes).map(x => adaptBottom(x._1, x._2))
+    val changeNeeded = tree.args == newArgs // cpy.Apply does not check if elements are the same,
```
I think this should be `tree.args != newArgs` or I'm misunderstanding what you're trying to do (if I'm right then you should add a test case for this)

smarter commented on src/dotty/tools/backend/jvm/BottomTypes.scala:
```diff
@@ -0,0 +1,81 @@
+    val newArgs: List[tpd.Tree] = tree.args.zip(fun.tpe.dealias.firstParamTypes).map(x => adaptBottom(x._1, x._2))
+    val changeNeeded = tree.args == newArgs // cpy.Apply does not check if elements are the same,
+                                            // it only does `eq` on lists as whole
+    if (changeNeeded) cpy.Apply(tree)(fun = fun, args = newArgs)
```
Why `fun = fun`?

DarkDimius replied:
this method does not have default arguments.

smarter commented on tests/pos/i828.scala:
```diff
@@ -0,0 +1,3 @@
+object X {
+  val x: Int = null.asInstanceOf[Nothing]
```
You should have test cases for the various nodes modified by `BottomTypes` and test cases for `Null` too.

smarter commented on src/dotty/tools/dotc/transform/Erasure.scala:
```diff
@@ -250,7 +250,7 @@ object Erasure extends TypeTestsCasts{
           adaptToType(tree.appliedToNone, pt)
         case tpw =>
-          if (pt.isInstanceOf[ProtoType] || tree.tpe <:< pt)
+          if (pt.isInstanceOf[ProtoType] || ((!pt.isValueType || !isNonJVMBottomType(tree.tpe)) && (tree.tpe <:< pt)))
```
The documentation of this method should be updated to explain what it does to bottom types too.

smarter commented:
`Nothing` is supposed to have no value, so I don't know if allowing `null.asInstanceOf[Nothing]` to run without any error is a good thing or not, maybe it doesn't matter.

odersky commented on src/dotty/tools/dotc/transform/Erasure.scala:
```diff
@@ -263,6 +263,12 @@ object Erasure extends TypeTestsCasts{
+    def isNonJVMBottomType(tpe: Type)(implicit ctx: Context): Boolean = {
+      tpe.derivesFrom(ctx.definitions.NothingClass) || tpe.derivesFrom(ctx.definitions.NullClass)
+    }
   }
```
Indeed, why not use Definitions#isBottomType?

DarkDimius replied:
Definitions#isBottomType(tp) is defined to be `tp.symbol \in {NothingClass, NullClass}` 
It does not work for types that do not return a usable symbol, such as And\Or types.

DarkDimius replied:
Should I update Definitions#isBottomType to account for `Null | Nothing` and `Int & Nothing`?

odersky replied:
No, I think it's OK to leave the two separate. Maybe rename the second to
`derivesFromBottomType`
to make clear that refinements and & / | count as well?


On Fri, Dec 18, 2015 at 6:32 PM, Dmitry Petrashko <notifications@github.com>
wrote:

> In src/dotty/tools/dotc/transform/Erasure.scala
> <https://github.com/lampepfl/dotty/pull/1002#discussion_r48048942>:
>
> > @@ -263,6 +263,12 @@ object Erasure extends TypeTestsCasts{
> >            else
> >              cast(tree, pt)
> >        }
> > +
> > +
> > +    /** Is `tpe` a type which is a bottom type for Dotty, but not a bottom type for JVM */
> > +    def isNonJVMBottomType(tpe: Type)(implicit ctx: Context): Boolean = {
> > +      tpe.derivesFrom(ctx.definitions.NothingClass) || tpe.derivesFrom(ctx.definitions.NullClass)
> > +    }
> >    }
>
> Should I update Definitions#isBottomType to account for Null | Nothing
> and Int & Nothing?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/lampepfl/dotty/pull/1002/files#r48048942>.
>



-- 
Martin Odersky
EPFL

DarkDimius replied:
Will do.

DarkDimius commented:
@smarter 
>Nothing is supposed to have no value, so I don't know if allowing null.asInstanceOf[Nothing] to run without any error is a good thing or not, maybe it doesn't matter.

I do not have a strong opinion on this either. So I've implemented `be compatible with scalac` mode.

```scala
class A {
 val s: Nothing  = null.asInstanceOf[Nothing]
}
```
emmits:
```java
public class A {
    private final Nothing. s = null;

    public Nothing. s() {
        return this.s;
    }
}
```

PS: the `.` in `Nothing.` is not a coincidence. It is a late transformation done by GenBCode.

smarter commented:
Being compatible with scalac seems OK here, but note that the following fails at runtime in `scalac` but works with your patch:
```scala
object X {
  def main(args: Array[String]): Unit = {
    val x: Int = null.asInstanceOf[Nothing]
  }
}
``

smarter commented:
And now that I think of it, this might be a good think because in the following:
```scala
object X {
  def main(args: Array[String]): Unit = {
    val x: Int = null.asInstanceOf
  }
}
```
The type parameter of `asInstanceOf` will be instantiated to `Nothing`, so it crashes at runtime in `scalac` but works with dotty.

odersky commented on src/dotty/tools/dotc/transform/Erasure.scala:
```diff
@@ -263,6 +263,12 @@ object Erasure extends TypeTestsCasts{
+    /** Is `tpe` a type which is a bottom type for Dotty, but not a bottom type for JVM */
+    def isNonJVMBottomType(tpe: Type)(implicit ctx: Context): Boolean = {
+      tpe.derivesFrom(ctx.definitions.NothingClass) || tpe.derivesFrom(ctx.definitions.NullClass)
+    }
```
I now did a similar change in definitions (f780a371a3716c6917a510ed72a5927704634bb3). 
So it's best if you cherry-pick the commit and use defn.isBottomType.

DarkDimius commented:
I've addressed most comments except the `(null.asInstanceOf[Nothing]): Int)` one.
I do not have a clear understanding of how this works in scalac and why is it different from the other case.
If you someone is interested to get to the bottom of this, It could be done in a different PR. This one fixes a compiler crash so I'd like to get this in.

sjrd commented:
FTR, in Scala.js we recently decided to go against what scalac/JVM does for `null.asInstanceOf[Nothing]`. Now it always throws immediately (as an undefined behavior). See the PR https://github.com/scala-js/scala-js/pull/2126 fixing issue https://github.com/scala-js/scala-js/issues/2125

DarkDimius commented:
/rebuild

odersky commented:
Let's either rebase and try again or close.
